### DISCUSSION TITLE
Async/Sync FastAPI

### DISCUSSION THREAD

**Discussion by Osbornnnnn** [fastapi/fastapi#6414] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Not need
```

### Description

FastAPI can be Async and Sync, but what type of database to use?
When using gunicorn and uvicorn, I can specify the number of Workers, for example 4.
How much will the speed of the application drop if I use a synchronous database?
Should Starlette create threads that simulate asynchrony?

### Operating System

Linux, Windows

### FastAPI Version

0.78.0

### Python Version

3.10.4

**Comment by JarroVGIT** [comment]
> How much will the speed of the application drop if I use a synchronous database?

That is hard to answer, without any context of your application. Is your application handling 1 request per hour or 500 per second? Using a synchronous database client would block your application, but it really depends on the application and its use case if that is a problem or not.

> Should Starlette create threads that simulate asynchrony?

Also hard to answer. My general rule of thumb: if you are in doubt, you probably need to go the async way.

**Comment by Kludex** [comment]
> FastAPI can be Async and Sync, but what type of database to use?

I don't know.

> When using gunicorn and uvicorn, I can specify the number of Workers, for example 4.

Reading the [Deployment section](https://fastapi.tiangolo.com/deployment/) should help.

> How much will the speed of the application drop if I use a synchronous database?

The wording here is not exactly correct... If you are looking for a benchmark, maybe this micro benchmark can help: https://github.com/tortoise/orm-benchmarks. It's not up-to-date, tho.

> Should Starlette create threads that simulate asynchrony?

I didn't get the question, but Starlette uses a `ThreadPoolExecutor` when the endpoint is _sync_.

**Comment by zoliknemet** [comment]
If you can use async database connection, use it with async endpoints!
If you have sync databasae connections, use it only with sync endpoints!

Avoid doing async endpoints with sync database connection because you could get some locks.
I will give you an example:


```python
async def get_sync_session():
	print("Fetching session")
	session = SyncSession()
    yield session
    session.close()


@app.get("/fetch", response_model=TestOutput)
async def fetch(session=Depends(get_sync_session)):
	print("Fetching data")
	response = session.fetch_something()
	print("Doing another IO operation")
	await asyncio.sleep()
	print("Returning response")
    return response

```

Lets say you have database connection pool of 2.
When you call `/fetch` third time and first 2 requests did not end it will go to lock.
Execution in loop:
request_1 Fetching session
request_1 Fetching data
request_1 Doing another IO operation # switch to request_2
request_2 Fetching session
request_2 Fetching data
request_2 Doing another IO operation # switch to request_3
request_3 Fetching session - LOCK happens here because it is not awaited

`get_sync_session` is executed in `event loop` and it does not goes to any await part and stays always in this state because `SyncSession()` will wait untill it gets connection. Since pool size is 2 there is no available connection and request_1 and request_2 do not get time in `event loop` to be executed to clear sessions `session.close()`.

It is specific combination, but you must be carefull.

My suggestion is to always use async db and async methods. Or use sync db and sync methods

In general avoid using ASYNC functions that do not have AWAIT part. Pyhton lets you do it, but those functions block `event loop` and if they are executed for long time nothing else gets executed until that function is over

NOTE: this is simplified explanation

**Comment by Osbornnnnn** [comment]
Thx for @zoliknemet.
I am testing async + async db, sync + sync db. DONT USE async + sunc db! )))

**Comment by zoliknemet** [comment]
@Osbornnnnn i recommend all async, read this: https://github.com/tiangolo/fastapi/issues/3205
there are still some problems id dependencies even in sync/sync...

**Comment by sandys** [comment]
i have both async and sync working with postgresql without lockups. i use the recommended contextmanager in sync to keep the coroutines in the same thread. 

https://gist.github.com/sandys/671b8b86ba913e6436d4cb22d04b135f
